traffic_control/clients/python/to_access/__init__.py

```diff
@@ -0,0 +1,576 @@
+
+.. option:: --request_headers
+
+	Output the request method line and any and all request headers. (Default: false)
+
+.. option:: --request_payload
+
+	Output the request body if any was sent. Will attempt to pretty-print the body as JSON if
+	:option:`--pretty` is used. (Default: false)
+
+.. option:: --response_headers
+
+	Output the response status line and any and all response headers. (Default: false)
+
+.. option:: --to_host HOST
+
+	The :abbr:`FQDN (Fully Qualified Domain Name)` and optionally the port and scheme of the Traffic
+	Ops server. This will override :envvar:`TO_URL`. The format is the same as for :envvar:`TO_URL`.
+	(Default: uses the value of :envvar:`TO_URL`)
+
+.. option:: --to_password PASSWORD
+
+	The password to use when authenticating to Traffic Ops. Overrides :envvar:`TO_PASSWORD`.
+	(Default: uses the value of :envvar:`TO_PASSWORD`)
+
+.. option:: --to_user USERNAME
+
+	The username to use when connecting to Traffic Ops. Overrides :envvar:`TO_USER`. (Default: uses
+	the value of :envvar:`TO_USER`)
+
+Environment Variables
+---------------------
+If defined, :program:`toaccess` scripts will use these environment variables to define their
+connection to and authentication with the Traffic Ops server. Typically, setting these is easier
+than using the long options :option:`--to_host`, :option:`--to_user`, and :option:`--to_password` on
+every invocation.
+
+.. envvar:: TO_PASSWORD
+
+	Will be used to authenticate the user defined by either :option:`--to_user` or :envvar:`TO_USER`.
+
+.. envvar:: TO_URL
+
+	The :abbr:`FQDN (Fully Qualified Domain Name)` of the Traffic Ops server to which the script
+	will connect. The format of this should be :file:`[{http or https}://]{hostname}[:{port}]`. Note
+	that this may optionally start with ``http://`` or ``https://`` (case insensitive), but
+	typically this is unnecessary. Also notice that the port number may be specified, though again
+	this isn't usually required. All :program:`toaccess` scripts will assume that port 443 should be
+	used unless otherwise specified. They will further assume that the protocol is HTTPS unless
+	:envvar:`TO_URL` (or :option:`--to_host`) starts with ``http://``, in which case the default port
+	will also be set to 80 unless otherwise specified in the URL.
+
+.. envvar:: TO_USER
+
+	The name of the user as whom to connect to the Traffic Ops server.
+
+Module Reference
+================
+
+"""
+import json
+import logging
+import os
+import sys
+import typing
+
+import requests
+
+from trafficops.restapi import LoginError, OperationError, InvalidJSONError
+from trafficops.tosession import TOSession
+
+l = logging.getLogger()
+l.disabled = True
+logging.basicConfig(level=logging.CRITICAL+1)
+
+#: The full path to a file used to store the user's Mojolicious athentication cookie (currently unused)
+COOKIEFILE = "" #os.path.expanduser(os.path.join("~", ".to-auth.cookie"))
+
+def setCookie(cookie:str):
```

Comment:
I always use caps for global constants, camelCase for functions and PascalCase for classes. It's still distinct. PEP8 sucks and I try to pretend it doesn't exist. Besides, it's meant as a convention for code included in the CPython reference implementation, and not as an advisement for other, unrelated projects. Finally, the `.pylintrc` file for our Python client checks for such formatting conventions, and doesn't place PEP8 restrictions on the codebase.

Reply:
PEP8 doesn't suck. Reading Python code that looks like a different language sucks. And it looks like a different language because the vast majority of python projects out there choose to follow PEP8 formatting convention. There's a reason projects choose to follow PEP8, and it's not because it's "good", it's because it's the biggest formatting convention and a best-practice to follow language conventions.

With this, we end up with snake_case in most places that use the standard library or virtually every other 3rd party library we import, and a mix of camelCase and snake_case in our own functions and variables, depending on who wrote the code. Why not just try to keep it all cohesive?

Reply:
>prefer pep8 formatting

+1

>PEP8 sucks

Obligatory Rob Pike quote:

>Gofmt's style is no one's favorite, yet gofmt is everyone's favorite.

Reply:
I think you're missing the point. I'm not mandating that we follow PEP8 100% in every single thing that it declares. It wasn't 100% PEP8 when it started, it most likely will never be 100% PEP8 in the future.

All I'm asking is that we stick to a conventional naming scheme, of which I think we should choose PEP8 because that allows the code to be consistent with the rest of the python standard library. Look through the python client code. It follows PEP8 recommended naming schemes. By choosing to use camelCase for new code, you are actually being inconsistent with the rest of the python client.

You also keep mentioning an existing Python style guide. This is a community of Python programmers, yet I don't recall seeing anything on the community mailing list about enforcing such a Python style guide. I for one would have things to say about such a Python style guide, as I'm sure others would too. You can't just reformat an entire project from space-indentation to tab-indentation, throw in a .pylintrc, and say that we have "an existing style guide". You essentially hijacked the existing informal-yet-not-codified style, changed it, and are now trying to enforce the changes without any community input.

Reply:
I'm not opposed to changing the naming scheme for the sake of consistency, but not if it means setting a precedence for conformance with a standard you yourself just said we don't follow.

The `pylintrc` was subject to the same review process as any other contribution, and is a living document subject to modification at will by the community. It's also not mandated anywhere as a standard, to be fair, and imo opinion shouldn't be until we drop Python2 support with the release of ATC 4.0 (hopefully before the end of this year) so that things like e.g. type annotations can be mandated without needing to worry about breaking Python2 support. If you have suggestions for formatting I'd encourage you to encode them in in the `pylintrc` so they can be enforced.

The upshot is: I'm happy to change formatting for the sake of consistency, but not to support a standard we aren't using, and currently have no plans to use.

Reply:
> I'm not opposed to changing the naming scheme for the sake of consistency, but not if it means setting a precedence for conformance with a standard you yourself just said we don't follow.

We can follow PEP8 naming conventions without conforming to the entire PEP8 standard itself. At the very least we should be using a consistent naming convention within the same package, but ideally we should follow the same convention across all python in TC.

> The pylintrc was subject to the same review process as any other contribution

We have a process. For things that would generally affect the broader community/project or bring in significant changes or new features, they should be discussed on the mailing list first to achieve some consensus. Things like a python style guide would affect anyone that contributes to or uses any python in Traffic Control and should therefore be discussed on the mailing list as opposed to a PR review.